crates/ty_python_semantic/resources/mdtest/narrow/conditionals/nested.md

```diff
@@ -238,6 +241,69 @@ def f(x: str | None):
 
         [reveal_type(x) for _ in range(1)]  # revealed: str
 
+    # When there is a reassignment, any narrowing constraints on the place are invalidated in lazy scopes.
+    x = None
+```
+
+If a variable defined in a private scope is never reassigned, narrowing remains in effect in the
+inner lazy scope.
+
+```py
+def f(const: str | None):
+    if const is not None:
+        def _():
+            # The `const is not None` narrowing constraint is still valid since `const` has not been reassigned
+            reveal_type(const)  # revealed: str
+
+        class C2:
+            reveal_type(const)  # revealed: str
+
+        [reveal_type(const) for _ in range(1)]  # revealed: str
+```
+
+And even if there is an attribute or subscript assignment to the variable, narrowing of the variable
+is still valid in the inner lazy scope.
+
+```py
+def f(l: list[str | None] | None):
+    if l is not None and l[0] is not None:
```

Comment:
`l[0] is not None` has no effect on this test, since we never check the type of `l[0]`

```suggestion
    if l is not None:
```